### ISSUE TITLE
(binary) not tokenized as T_BINARY_CAST

### ISSUE THREAD

**Issue by jrfnl** [squizlabs/PHP_CodeSniffer#1574]
There are two ways to cast to binary as introduced in PHP 5.2.1:
* `$binary = b"binary string";`
* `$binary = (binary) $string;`

The first one is identified by PHPCS as `T_BINARY_CAST`, while the second is not.

By the looks of it, PHP itself incorrectly tokenizes this as `T_STRING_CAST`. See: https://3v4l.org/n0DJj

Ref: http://php.net/manual/en/language.types.type-juggling.php#language.types.typecasting

**Comment by jrfnl** [comment]
On second look: neither are currently tokenized correctly.

It does look like there is some code in the PHP Tokenizer which is supported to take care of the first case, but by the looks of it, it's not working correctly.

See: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Tokenizers/PHP.php#L567-L575

**Comment by gsherwood** [comment]
The binary cast code only gets called if PHP has tokenized the contents of the double quoted string. So it needs some variable in it, or some other reason for PHP to break it up, such as: `$binary = b"binary string $foo";`

I can fix that, and convert the `(binary)` cast into the T_BINARY_CAST token I created for the `b` type cast despite PHP calling it a T_STRING_CAST, which it presumably does because it is casting to a "binary string". Feels like it makes more sense.

**Comment by jrfnl** [comment]
> The binary cast code only gets called if PHP has tokenized the contents of the double quoted string. So it needs some variable in it, or some other reason for PHP to break it up, such as: $binary = b"binary string $foo";

I didn't know that and the example in the PHP documentation doesn't show that either. Is the variable a PHPCS requirement for the token parsing or a PHP requirement for binary casting ?

> I can fix that, and convert the (binary) cast into the T_BINARY_CAST token I created for the b type cast despite PHP calling it a T_STRING_CAST, which it presumably does because it is casting to a "binary string". Feels like it makes more sense.

Well, they both make sense in a way, but as all the casts have their own token, it felt a bit odd that this one doesn't.

For the sniff where I noticed the issue, I've worked around it for now, but I can imagine this may be useful/more intuitive to work with for people anyway

**Comment by gsherwood** [comment]
> I didn't know that and the example in the PHP documentation doesn't show that either. Is the variable a PHPCS requirement for the token parsing or a PHP requirement for binary casting ?

Sorry, I forgot to reply to this.

It's not a PHP thing - it's a PHPCS thing. PHPCS is only doing that extra binary cast detection when the string has a variable in it. This is because PHP tokenizes the string token in a different way and PHPCS is only looking for that one type of token structure. It should be changed to look for both.

**Comment by jrfnl** [comment]
Thanks for your reply. Sounds like a plan.

**Comment by jrfnl** [comment]
Just ran into this issue again. I look forward to seeing a solution at some point.

**Comment by gsherwood** [comment]
I've switched all types (I hope) of binary casting over to the T_BINARY_CAST token now.

**Comment by jrfnl** [comment]
@gsherwood Much appreciated! I'll go and test and if necessary adjust various (external) sniffs to deal with the update ;-)

**Comment by jrfnl** [comment]
Uh oh... looks like the tokenizer does not handle casts with whitespace in them correctly, i.e. the below are still tokenized as `T_STRING_CAST`:
```php
(   binary  ) 1234;
( BINARY ) 1234;
```

**Comment by jrfnl** [comment]
PR #2215 should take care of the above.
